### PR TITLE
Check stock before doing anything related to admin payments

### LINKED ISSUE #7614: Multiple Stripe payments collected for incomplete order created in back office
## Description

Order ID R506827701

Order created in the back office by hub manager.
Each time they added a card payment it stated that there was 'An Error Saving Payment'
imbywallpaymentbackend

I am unable to find a bugsnag for this.  NOTE billing name is different to the name on bank card used for final payment, but consistent otherwise

Order remains 'incomplete' at 'Payment' stage on OFN platform with four pending payments showing:
imbywallorder

Stripe records show that all four payments have been collected:
imbywallstripe

Clicking on the 'X' to void the last three payments on the OFN system and 'tick' on the first (to capture one payment on OFN and hence move the order to 'complete' for delivery and supplier reporting) resulted in three payments as 'void' and one as 'failed'. 
imbywallorderpost
But on Stripe ALL FOUR are refunded:#
imbywallstripepost

## Expected Behavior

Order created in back office.
Shopper pays by card over and hub manager enters in card details.
If card details are correct, OFN and Stripe both record payment has successfully been taken.  OFN order moves from the incomplete 'Payment' status to 'complete' 'paid' status.
If card details are not correct OFN shows incomplete payment failed.  No payment collected by Stripe- this also shows 'failed' payment attempt.  Order remains at 'Payment' status on OFN platform.

## Actual Behaviour

Order created in back office.
Shopper pays by card over and hub manager enters in card details.
OFN displays 'Error in taking payment' notice but Stripe processes the payment.
Order remains in 'Payment' incomplete status on OFN platform.
Payment collected from shopper multiple times.

## Steps to Reproduce

1.
2.
3.
4.

## Severity

Bug S1

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by andrewpbrett**
#### What? Why?

Closes #7614 

We could go about fixing this in a few different ways. The lightest, most specific way is to explicitly check stock in a before action in the controller. If there is insufficient stock, and the order is in the payment state, we alert the admin and redirect to the order details page. 

This is also, however, a fantastic example of a place where we should think about unifying the order and payment state transitions between front and backend, as in https://github.com/openfoodfoundation/openfoodnetwork/issues/7511. Ping @lin-d-hop & @filipefurtad0. And an example of how the transitions in the two models are coupled pretty tightly. 

#### What should we test?

1. create admin order
2. add product that is in stock
3. enter customer details
4. confirm that order is now in Payment state
5. set product to 0 stock
6. visit the Payments screen on the admin order page
7. it should redirect to the order details page and have a flash message saying that the product is out of stock. 

#### Release notes

Fixed a bug where creating an admin payment for an order with out of stock items returned an error message but created the payment in Stripe. 

Changelog Category: User facing changes

### CODE REVIEW

File: app/controllers/spree/admin/payments_controller.rb [hunk 1]
```diff
@@ -142,6 +143,17 @@ def can_transition_to_payment
         redirect_to spree.edit_admin_order_customer_url(@order)
       end
 
+      def ensure_sufficient_stock_lines
+        return if @order.insufficient_stock_lines.blank? || !@order.payment?
```
Comment (Matt-Yorkley) [inline app/controllers/spree/admin/payments_controller.rb:147]: Great.

I'm wondering if `!@order.payment?` here should be `@order.competed?` instead?

Also; with these two parts of the conditional; the first part (stock check with variant override scoping) is expensive and the second part is really cheap. If we switch the order, we can fail early in lots of cases and avoid running the stock check entirely, eg:
```ruby
return if @order.competed? || @order.insufficient_stock_lines.blank? 
```

File: app/controllers/spree/admin/payments_controller.rb [hunk 2]
```diff
@@ -9,6 +9,7 @@ class PaymentsController < Spree::Admin::BaseController
       before_action :load_payment, only: [:fire, :show]
       before_action :load_data
       before_action :can_transition_to_payment
+      before_action :ensure_sufficient_stock_lines
```
Comment (Matt-Yorkley) [inline app/controllers/spree/admin/payments_controller.rb:12]: Do we want to run this on al of this controller's actions? Is it really needed on index, for example? I'm wondering if this should be scoped to specific actions, like:
```ruby
      before_action :ensure_sufficient_stock_lines, only: [:create, :update]
```
:thinking:

**Comment by andrewpbrett** [comment]
> I'm wondering if !@order.payment? here should be @order.competed? instead?
> Do we want to run this on all of the controller's actions?

@Matt-Yorkley I had the exact same thought process while writing this, and in fact had it written that way and then changed it 😂. I should have mentioned it in the description. 

I think we do want to prevent the screen from showing up at all. If we load the screen, we're letting someone tediously manually enter a card number when we already know, for sure, that we're about to throw an error and kick them back to the order details screen. 

Therefore, in the case there `@order.completed == true`, we *do* want to allow access to the payments/index page, so that an admin can view the successful payment. 

That is an excellent catch on switching the order of the conditional, though, I'll make that change.

File: app/controllers/spree/admin/payments_controller.rb (hunk 1 above)
Comment (andrewpbrett) [reply app/controllers/spree/admin/payments_controller.rb:147]:
```suggestion
        return if !@order.payment? || @order.insufficient_stock_lines.blank?
```

File: app/controllers/spree/admin/payments_controller.rb [hunk 3]
```diff
@@ -142,6 +143,17 @@ def can_transition_to_payment
...
+        return if !@order.payment? || @order.insufficient_stock_lines.blank?
+
+        out_of_stock_items = @order.insufficient_stock_lines.map do |line_item|
+          line_item.variant.name
+        end.join(", ")
+        flash[:error] = I18n.t("spree.orders.line_item.insufficient_stock",
+          on_hand: "0 #{out_of_stock_items}")
+        redirect_to spree.edit_admin_order_url(@order)
+      end
+
```
Comment (luisramos0) [inline app/controllers/spree/admin/payments_controller.rb:156]: This week I have been teaching refactoring techniques where "Extract Method" is king.
That out_of_stock_items is a wonderful candidate to be extracted as a method, no params required :ok_hand:

**Review by luisramos0 — APPROVED** [review]
Nice!

Based on your explanation Andy I think that line
before_action :ensure_sufficient_stock_lines
needs a short comment explaining why it's active for all actions.

**Review by sauloperez — APPROVED** [review]
neat :ok_hand: pretty good exercise of finding out the true root cause :clap:

**Comment by filipefurtad0** [comment]
Hey @andrewpbrett ,

Thanks for the detailed notes on what to test - it is indeed possible to reproduce the bug, by following those steps,

**before staging the PR**

1) if a **stripe 3D-auth card is used**, and authentication is completed:

image

Notice:
- the :cat: (422) 
- the payment state remains the same, although...
- ...the payment is taken on Stripe, as seen on the dashboard

2) if a **non-3D auth card** is used:

image

Notice:
- the "Error saving payment"
- The amount being captured on Stripe


3) **cash is used as payment method**

image

Triggers:
https://app.bugsnag.com/yaycode/openfoodnetwork/errors/601f3ded27328e0017d8a779?event_id=60a54313007ca63488ef0000&i=sk&m=nw


**After this PR:**

1) if a **stripe 3D-auth card is used**, and authentication is completed:

There are two scenarios here worth considering:

i) the stock is reduced _before_ the Stripe payment is added (and authorization is provided by the customer)

We're good on this one: the error message "Insufficient stock available, only ... stock limit product remaining." :tada: 
image


ii) the stock is reduced _after_ the Stripe payment is added, but _before_ authorization is provided by the customer

In this case, we still get the 422, And the stripe payment still goes through, as seen on the Stripe-Dashboard:

image

2) non-3D stripe card
The message "Insufficient stock available, only 0 stock limit product remaining":

image

- the order remains in the state "Payment"
- no payment is shown on Stripe Dashboard, for this order

3) cash payment
The user sees the message:
image


**TL;DR** - So, we're good on all cases, except 1.ii) in which the stock is reduced somewhere between the moment the admin creates the payment till the moment the customer provides authorization.

I think we should merge this PR as it is a huge improvement :clap: :clap: :clap:  and covers all other cases. I'll add an issue regarding this last case - I think it should be an S2, as the time between payment authorization from customer may be considerable.

Moving to ready to go.
